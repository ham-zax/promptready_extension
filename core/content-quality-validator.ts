// Content Quality Validator - ensures high-quality markdown output
// Provides comprehensive validation and quality scoring

export interface QualityMetrics {
  contentPreservation: number;
  structureIntegrity: number;
  markdownQuality: number;
  readability: number;
  completeness: number;
}

export interface QualityReport {
  overallScore: number;
  metrics: QualityMetrics;
  issues: QualityIssue[];
  recommendations: string[];
  passesThreshold: boolean;
}

export interface QualityIssue {
  type: 'error' | 'warning' | 'info';
  category: 'content' | 'structure' | 'formatting' | 'metadata';
  message: string;
  severity: number; // 1-10, 10 being most severe
  suggestion?: string;
}

export interface ValidationOptions {
  minContentLength: number;
  maxContentReduction: number; // Maximum allowed content reduction (0-1)
  minStructurePreservation: number; // Minimum structure preservation ratio (0-1)
  qualityThreshold: number; // Minimum overall quality score (0-100)
  strictMode: boolean;
}

export class ContentQualityValidator {
  
  private static readonly DEFAULT_OPTIONS: ValidationOptions = {
    minContentLength: 500,
    maxContentReduction: 0.9, // Allow up to 90% reduction
    minStructurePreservation: 0.5, // Require at least 50% structure preservation
    qualityThreshold: 70, // Minimum score of 70/100
    strictMode: false,
  };

  /**
   * Validate processed markdown content quality
   */
  static validate(
    markdown: string,
    originalHtml: string,
    processingStats: any,
    options: Partial<ValidationOptions> = {}
  ): QualityReport {
    const config = { ...this.DEFAULT_OPTIONS, ...options };
    const issues: QualityIssue[] = [];
    const recommendations: string[] = [];

    console.log('[ContentQualityValidator] Starting quality validation...');

    // Calculate individual metrics
    const contentPreservation = this.calculateContentPreservation(markdown, originalHtml, issues);
    const structureIntegrity = this.calculateStructureIntegrity(markdown, originalHtml, issues);
    const markdownQuality = this.calculateMarkdownQuality(markdown, issues);
    const readability = this.calculateReadability(markdown, issues);
    const completeness = this.calculateCompleteness(markdown, originalHtml, processingStats, issues);

    const metrics: QualityMetrics = {
      contentPreservation,
      structureIntegrity,
      markdownQuality,
      readability,
      completeness,
    };

    // Calculate overall score (weighted average)
    const weights = {
      contentPreservation: 0.25,
      structureIntegrity: 0.20,
      markdownQuality: 0.20,
      readability: 0.15,
      completeness: 0.20,
    };

    const overallScore = Object.entries(metrics).reduce((score, [metric, value]) => {
      return score + (value * weights[metric as keyof QualityMetrics]);
    }, 0);

    // Generate recommendations
    this.generateRecommendations(metrics, issues, recommendations, config);

    // Check if passes threshold
    const passesThreshold = overallScore >= config.qualityThreshold && 
                           this.meetsMinimumRequirements(metrics, config);

    console.log(`[ContentQualityValidator] Validation complete. Score: ${overallScore.toFixed(1)}/100`);

    return {
      overallScore: Math.round(overallScore * 10) / 10,
      metrics,
      issues: issues.sort((a, b) => b.severity - a.severity),
      recommendations,
      passesThreshold,
    };
  }

  /**
   * Calculate content preservation score
   */
  private static calculateContentPreservation(
    markdown: string,
    originalHtml: string,
    issues: QualityIssue[]
  ): number {
    const markdownLength = markdown.length;
    const originalLength = originalHtml.length;
    
    if (originalLength === 0) {
      issues.push({
        type: 'error',
        category: 'content',
        message: 'Original content is empty',
        severity: 10,
      });
      return 0;
    }

    const reductionRatio = 1 - (markdownLength / originalLength);
    
    // Score based on content reduction
    let score = 100;
    
    if (reductionRatio > 0.95) {
      score = 10;
      issues.push({
        type: 'error',
        category: 'content',
        message: 'Excessive content loss (>95%)',
        severity: 9,
        suggestion: 'Check Readability configuration and fallback mechanisms',
      });
    } else if (reductionRatio > 0.9) {
      score = 30;
      issues.push({
        type: 'warning',
        category: 'content',
        message: 'High content loss (>90%)',
        severity: 7,
        suggestion: 'Consider adjusting content extraction parameters',
      });
    } else if (reductionRatio > 0.8) {
      score = 60;
      issues.push({
        type: 'warning',
        category: 'content',
        message: 'Moderate content loss (>80%)',
        severity: 5,
      });
    } else if (reductionRatio < 0.1) {
      score = 70;
      issues.push({
        type: 'info',
        category: 'content',
        message: 'Minimal content filtering (<10% reduction)',
        severity: 3,
        suggestion: 'Content may contain noise that could be filtered',
      });
    }

    // Check for minimum content length
    if (markdownLength < 200) {
      score = Math.min(score, 20);
      issues.push({
        type: 'error',
        category: 'content',
        message: 'Output content too short',
        severity: 8,
      });
    }

    return score;
  }

  /**
   * Calculate structure integrity score
   */
  private static calculateStructureIntegrity(
    markdown: string,
    originalHtml: string,
    issues: QualityIssue[]
  ): number {
    let score = 100;

    // Count structural elements
    const markdownHeadings = (markdown.match(/^#{1,6}\s/gm) || []).length;
    const originalHeadings = (originalHtml.match(/<h[1-6][^>]*>/gi) || []).length;
    
    const markdownLists = (markdown.match(/^[\s]*[-*+]\s/gm) || []).length;
    const originalLists = (originalHtml.match(/<[uo]l[^>]*>/gi) || []).length;
    
    const markdownCodeBlocks = (markdown.match(/```[\s\S]*?```/g) || []).length;
    const originalCodeBlocks = (originalHtml.match(/<pre[^>]*>/gi) || []).length;

    // Check heading preservation
    if (originalHeadings > 0) {
      const headingPreservation = markdownHeadings / originalHeadings;
      if (headingPreservation < 0.3) {
        score -= 30;
        issues.push({
          type: 'error',
          category: 'structure',
          message: 'Poor heading preservation (<30%)',
          severity: 8,
          suggestion: 'Check heading extraction rules',
        });
      } else if (headingPreservation < 0.6) {
        score -= 15;
        issues.push({
          type: 'warning',
          category: 'structure',
          message: 'Moderate heading loss',
          severity: 5,
        });
      }
    }

    // Check list preservation
    if (originalLists > 0) {
      const listPreservation = markdownLists / originalLists;
      if (listPreservation < 0.5) {
        score -= 20;
        issues.push({
          type: 'warning',
          category: 'structure',
          message: 'Poor list preservation',
          severity: 6,
        });
      }
    }

    // Check code block preservation
    if (originalCodeBlocks > 0) {
      const codePreservation = markdownCodeBlocks / originalCodeBlocks;
      if (codePreservation < 0.7) {
        score -= 25;
        issues.push({
          type: 'error',
          category: 'structure',
          message: 'Poor code block preservation',
          severity: 7,
          suggestion: 'Verify code block extraction and conversion rules',
        });
      }
    }

    return Math.max(0, score);
  }

  /**
   * Calculate markdown quality score
   */
  private static calculateMarkdownQuality(
    markdown: string,
    issues: QualityIssue[]
  ): number {
    let score = 100;

    // Check for HTML tags in markdown
    const htmlTagMatches = markdown.match(/<[^>]+>/g);
    if (htmlTagMatches && htmlTagMatches.length > 0) {
      const htmlTagCount = htmlTagMatches.length;
      score -= Math.min(40, htmlTagCount * 2);
      issues.push({
        type: 'warning',
        category: 'formatting',
        message: `HTML tags found in markdown output (${htmlTagCount})`,
        severity: 6,
        suggestion: 'Improve HTML to markdown conversion rules',
      });
    }

    // Check for excessive whitespace
    const excessiveWhitespace = markdown.match(/\n{4,}/g);
    if (excessiveWhitespace && excessiveWhitespace.length > 0) {
      score -= 10;
      issues.push({
        type: 'info',
        category: 'formatting',
        message: 'Excessive whitespace detected',
        severity: 3,
        suggestion: 'Enable post-processing whitespace cleanup',
      });
    }

    // Check for malformed links
    const malformedLinks = markdown.match(/\[([^\]]*)\]\(\s*\)/g);
    if (malformedLinks && malformedLinks.length > 0) {
      score -= 15;
      issues.push({
        type: 'warning',
        category: 'formatting',
        message: 'Malformed links detected',
        severity: 5,
        suggestion: 'Improve link processing and validation',
      });
    }

    // Check for proper heading hierarchy
    const headings = markdown.match(/^(#{1,6})\s/gm);
    if (headings && headings.length > 1) {
      const levels = headings.map(h => h.length - 1);
      let hierarchyIssues = 0;
      
      for (let i = 1; i < levels.length; i++) {
        if (levels[i] > levels[i-1] + 1) {
          hierarchyIssues++;
        }
      }
      
      if (hierarchyIssues > 0) {
        score -= hierarchyIssues * 5;
        issues.push({
          type: 'info',
          category: 'structure',
          message: 'Heading hierarchy issues detected',
          severity: 4,
          suggestion: 'Enable heading normalization in post-processing',
        });
      }
    }

    return Math.max(0, score);
  }

  /**
   * Calculate readability score
   */
  private static calculateReadability(
    markdown: string,
    issues: QualityIssue[]
  ): number {
    let score = 100;

    // Calculate basic readability metrics
    const sentences = markdown.split(/[.!?]+/).filter(s => s.trim().length > 0);
    const words = markdown.split(/\s+/).filter(w => w.length > 0);
    const paragraphs = markdown.split(/\n\s*\n/).filter(p => p.trim().length > 0);

    if (sentences.length === 0 || words.length === 0) {
      score = 0;
      issues.push({
        type: 'error',
        category: 'content',
        message: 'No readable content found',
        severity: 10,
      });
      return score;
    }

    // Average sentence length
    const avgSentenceLength = words.length / sentences.length;
    if (avgSentenceLength > 30) {
      score -= 15;
      issues.push({
        type: 'info',
        category: 'content',
        message: 'Long average sentence length may affect readability',
        severity: 3,
      });
    }

    // Paragraph structure
    if (paragraphs.length < 2 && words.length > 100) {
      score -= 20;
      issues.push({
        type: 'warning',
        category: 'structure',
        message: 'Poor paragraph structure',
        severity: 5,
        suggestion: 'Content may need better paragraph breaks',
      });
    }

    // Check for very short content
    if (words.length < 50) {
      score -= 30;
      issues.push({
        type: 'warning',
        category: 'content',
        message: 'Very short content may not be meaningful',
        severity: 6,
      });
    }

    return Math.max(0, score);
  }

  /**
   * Calculate completeness score
   */
  private static calculateCompleteness(
    markdown: string,
    originalHtml: string,
    processingStats: any,
    issues: QualityIssue[]
  ): number {
    let score = 100;

    // Check if processing used fallbacks
    if (processingStats?.fallbacksUsed?.length > 0) {
      const fallbackPenalty = processingStats.fallbacksUsed.length * 15;
      score -= fallbackPenalty;
      issues.push({
        type: 'warning',
        category: 'content',
        message: `Processing used ${processingStats.fallbacksUsed.length} fallback(s)`,
        severity: 6,
        suggestion: 'Primary processing methods failed, check configuration',
      });
    }

    // Check processing time (very fast might indicate failure)
    if (processingStats?.totalTime < 100) {
      score -= 10;
      issues.push({
        type: 'info',
        category: 'metadata',
        message: 'Very fast processing time may indicate incomplete processing',
        severity: 3,
      });
    }

    // Check for processing errors
    if (processingStats?.errors?.length > 0) {
      score -= 40;
      issues.push({
        type: 'error',
        category: 'content',
        message: 'Processing errors occurred',
        severity: 8,
      });
    }

    return Math.max(0, score);
  }

  /**
   * Generate recommendations based on metrics and issues
   */
  private static generateRecommendations(
    metrics: QualityMetrics,
    issues: QualityIssue[],
    recommendations: string[],
    config: ValidationOptions
  ): void {
    // Content preservation recommendations
    if (metrics.contentPreservation < 60) {
      recommendations.push('Consider adjusting Readability configuration for better content extraction');
      recommendations.push('Review content extraction selectors and rules');
    }

    // Structure integrity recommendations
    if (metrics.structureIntegrity < 70) {
      recommendations.push('Enable structure-preserving options in Turndown configuration');
      recommendations.push('Review heading and list conversion rules');
    }

    // Markdown quality recommendations
    if (metrics.markdownQuality < 80) {
      recommendations.push('Enable post-processing to clean up markdown formatting');
      recommendations.push('Review Turndown custom rules for better conversion');
    }

    // Readability recommendations
    if (metrics.readability < 70) {
      recommendations.push('Consider content preprocessing to improve structure');
      recommendations.push('Enable paragraph normalization in post-processing');
    }

    // Completeness recommendations
    if (metrics.completeness < 80) {
      recommendations.push('Review processing pipeline for potential failures');
      recommendations.push('Consider enabling more robust fallback mechanisms');
    }

    // High-severity issue recommendations
    const highSeverityIssues = issues.filter(issue => issue.severity >= 7);
    if (highSeverityIssues.length > 0) {
      recommendations.push('Address high-severity issues to improve overall quality');
    }

    // Remove duplicates
    const uniqueRecommendations = [...new Set(recommendations)];
    recommendations.splice(0, recommendations.length, ...uniqueRecommendations);
  }

  /**
   * Check if metrics meet minimum requirements
   */
  private static meetsMinimumRequirements(
    metrics: QualityMetrics,
    config: ValidationOptions
  ): boolean {
    if (config.strictMode) {
      return Object.values(metrics).every(score => score >= 70);
    }

    // At least content preservation and structure integrity should be decent
    return metrics.contentPreservation >= 50 && metrics.structureIntegrity >= 40;
  }

  /**
   * Get quality grade based on score
   */
  static getQualityGrade(score: number): { grade: string; description: string } {
    if (score >= 90) {
      return { grade: 'A', description: 'Excellent quality' };
    } else if (score >= 80) {
      return { grade: 'B', description: 'Good quality' };
    } else if (score >= 70) {
      return { grade: 'C', description: 'Acceptable quality' };
    } else if (score >= 60) {
      return { grade: 'D', description: 'Poor quality' };
    } else {
      return { grade: 'F', description: 'Failed quality check' };
    }
  }

  /**
   * Generate quality summary report
   */
  static generateSummaryReport(report: QualityReport): string {
    const grade = this.getQualityGrade(report.overallScore);
    const errorCount = report.issues.filter(i => i.type === 'error').length;
    const warningCount = report.issues.filter(i => i.type === 'warning').length;

    return `Quality Report: ${grade.grade} (${report.overallScore}/100)
${grade.description}

Issues: ${errorCount} errors, ${warningCount} warnings
${report.passesThreshold ? '✅ Passes quality threshold' : '❌ Below quality threshold'}

Top Recommendations:
${report.recommendations.slice(0, 3).map(r => `• ${r}`).join('\n')}`;
  }
}
