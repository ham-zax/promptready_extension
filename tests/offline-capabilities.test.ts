// Comprehensive test suite for the offline capabilities system
// Tests all components and integration scenarios

import { describe, test, expect } from 'vitest';
import { OfflineModeManager } from '../core/offline-mode-manager';
import { ReadabilityConfigManager } from '../core/readability-config';
import { TurndownConfigManager } from '../core/turndown-config';
import { MarkdownPostProcessor } from '../core/post-processor';
import { ContentQualityValidator } from '../core/content-quality-validator';
import { ErrorHandler } from '../core/error-handler';

// Mock DOM environment for testing
const mockDOMParser = {
  parseFromString: (html: string, type: string) => {
    // Simple mock implementation
    return {
      body: {
        innerHTML: html,
        textContent: html.replace(/<[^>]*>/g, ''),
        children: [],
        querySelectorAll: () => [],
        querySelector: () => null,
      },
      title: 'Test Document',
      cloneNode: () => mockDOMParser.parseFromString(html, type),
    };
  },
};

// @ts-ignore
global.DOMParser = function() {
  return mockDOMParser;
};

describe('Offline Capabilities System', () => {
  
  describe('OfflineModeManager', () => {
    
    test('should process simple HTML content successfully', async () => {
      const html = `
        <html>
          <body>
            <h1>Test Article</h1>
            <p>This is a test paragraph with some content.</p>
            <p>Another paragraph with more content.</p>
          </body>
        </html>
      `;
      
      const result = await OfflineModeManager.processContent(
        html,
        'https://example.com/test',
        'Test Article'
      );
      
      expect(result.success).toBe(true);
      expect(result.markdown).toContain('# Test Article');
      expect(result.markdown).toContain('test paragraph');
      expect(result.processingStats.totalTime).toBeGreaterThan(0);
      expect(result.warnings).toBeDefined();
      expect(result.errors).toHaveLength(0);
    });
    
    test('should handle large content with chunking', async () => {
      const largeContent = '<article><h1>Large</h1><p>' + 'Large content '.repeat(50000) + '</p></article>';
      const html = `<html><body>${largeContent}</body></html>`;

      const result = await OfflineModeManager.processLargeContent(
        html,
        'https://example.com/large',
        'Large Document'
      );

      expect(result.success).toBe(true);
      expect(typeof result.markdown).toBe('string');
      expect(result.processingStats.fallbacksUsed).toBeDefined();
    });
    
    test('should get optimal configuration for different URLs', async () => {
      const githubConfig = await OfflineModeManager.getOptimalConfig('https://github.com/user/repo', { renderer: 'unified' });
      expect(githubConfig.readabilityPreset).toBe('technical-documentation');
      expect(githubConfig.turndownPreset).toBe('github');

      const blogConfig = await OfflineModeManager.getOptimalConfig('https://blog.example.com/post', { renderer: 'unified' });
      expect(blogConfig.readabilityPreset).toBe('blog-article');
      expect(blogConfig.postProcessing.addTableOfContents).toBe(true);
    });
    
  });
  
  describe('ReadabilityConfigManager', () => {
    
    test('should detect content type from URL patterns', () => {
      const techConfig = ReadabilityConfigManager.getConfigForUrl('https://docs.example.com/api');
      expect(techConfig.charThreshold).toBe(300); // Technical docs threshold

      const newsConfig = ReadabilityConfigManager.getConfigForUrl('https://news.example.com/2024/01/01/article');
      expect(newsConfig.charThreshold).toBe(800); // News article threshold (updated)
    });
    
    test('should provide fallback configuration for unknown URLs', () => {
      const defaultConfig = ReadabilityConfigManager.getConfigForUrl('https://unknown-site.com');
      expect(defaultConfig.charThreshold).toBe(500); // Default threshold
      expect(defaultConfig.classesToPreserve).toContain('highlight');
    });
    
    test('should validate content quality', () => {
      const content = '<p>Good content</p>'.repeat(100);
      const originalLength = 10000;
      
      const validation = ReadabilityConfigManager.validateContentQuality(content, originalLength);
      expect(validation.isValid).toBe(true);
      expect(validation.score).toBeGreaterThan(60);
    });
    
  });
  
  describe('TurndownConfigManager', () => {
    
    test('should convert HTML to markdown with different presets', async () => {
      const html = `
        <h1>Title</h1>
        <p>Paragraph with <strong>bold</strong> and <em>italic</em> text.</p>
        <pre><code class="language-javascript">console.log('hello');</code></pre>
        <ul><li>Item 1</li><li>Item 2</li></ul>
      `;
      
      const standardMarkdown = await TurndownConfigManager.convert(html, 'standard');
      expect(standardMarkdown).toContain('# Title');
      expect(standardMarkdown).toContain('**bold**');
      expect(standardMarkdown).toContain('*italic*');
      expect(standardMarkdown).toContain('```javascript');
      
      const githubMarkdown = await TurndownConfigManager.convert(html, 'github');
      expect(githubMarkdown).toContain('```javascript');
      
      const obsidianMarkdown = await TurndownConfigManager.convert(html, 'obsidian');
      expect(obsidianMarkdown).toContain('# Title');
    });
    
    test('should validate markdown output quality', () => {
      const markdown = '# Title\n\nParagraph with **bold** text.\n\n```js\nconsole.log("test");\n```';
      const originalHtml = '<h1>Title</h1><p>Paragraph with <strong>bold</strong> text.</p><pre><code>console.log("test");</code></pre>';
      
      const validation = TurndownConfigManager.validateMarkdown(markdown, originalHtml);
      expect(validation.isValid).toBe(true);
      expect(validation.score).toBeGreaterThan(70);
      expect(validation.stats.headings).toBe(1);
      expect(validation.stats.codeBlocks).toBe(1);
    });
    
  });
  
  describe('MarkdownPostProcessor', () => {
    
    test('should clean up markdown formatting', () => {
      const messyMarkdown = `
        #  Title



        Paragraph with   trailing spaces



        Another paragraph




      `;

      const result = MarkdownPostProcessor.process(messyMarkdown);
      const body = result.markdown.replace(/```[\s\S]*?```/g, '');
      expect(body).not.toContain('   '); // No trailing spaces outside code blocks
      expect(result.markdown).not.toMatch(/\n{3,}/); // No excessive newlines
      expect(result.improvements.length).toBeGreaterThan(0);
    });
    
    test('should normalize heading structure', () => {
      const badHeadings = '# Title\n\n#### Skipped Level\n\n## Proper Level';
      
      const result = MarkdownPostProcessor.process(badHeadings, {
        normalizeHeadings: true,
        cleanupWhitespace: true,
        removeEmptyLines: true,
        maxConsecutiveNewlines: 2,
        fixListFormatting: false,
        improveCodeBlocks: false,
        enhanceLinks: false,
        optimizeImages: false,
        addTableOfContents: false,
        preserveLineBreaks: false,
      });
      
      expect(result.improvements).toContain('Fixed heading hierarchy');
    });
    
    test('should add table of contents when requested', () => {
      const markdown = '# Main Title\n\n## Section 1\n\n### Subsection\n\n## Section 2';
      
      const result = MarkdownPostProcessor.process(markdown, {
        addTableOfContents: true,
        cleanupWhitespace: true,
        normalizeHeadings: true,
        removeEmptyLines: true,
        maxConsecutiveNewlines: 2,
        fixListFormatting: false,
        improveCodeBlocks: false,
        enhanceLinks: false,
        optimizeImages: false,
        preserveLineBreaks: false,
      });
      
      expect(result.markdown).toContain('## Table of Contents');
      expect(result.markdown).toContain('- [Section 1]');
    });
    
  });
  
  describe('ContentQualityValidator', () => {
    
    test('should validate high-quality content', () => {
      const goodMarkdown = `
        # Article Title
        
        This is a well-structured article with multiple paragraphs.
        
        ## Section 1
        
        Content with proper structure and formatting.
        
        - List item 1
        - List item 2
        
        \`\`\`javascript
        console.log('code example');
        \`\`\`
        
        ## Section 2
        
        More content with [links](https://example.com) and **formatting**.
      `;
      
      const originalHtml = '<h1>Article Title</h1><p>Content...</p>'.repeat(10);
      const mockStats = {
        totalTime: 1000,
        fallbacksUsed: [],
        errors: [],
      };
      
      const report = ContentQualityValidator.validate(goodMarkdown, originalHtml, mockStats);
      expect(report.overallScore).toBeGreaterThan(70);
      expect(report.passesThreshold).toBe(true);
      expect(report.metrics.markdownQuality).toBeGreaterThan(80);
    });
    
    test('should detect poor-quality content', () => {
      const poorMarkdown = 'Short content';
      const originalHtml = '<div>Much longer original content</div>'.repeat(100);
      const mockStats = {
        totalTime: 100,
        fallbacksUsed: ['readability-fallback', 'turndown-fallback'],
        errors: ['Processing error'],
      };

      const report = ContentQualityValidator.validate(poorMarkdown, originalHtml, mockStats);
      expect(report.overallScore).toBeLessThan(60);
      expect(report.passesThreshold).toBe(false);
      expect(report.issues.length).toBeGreaterThan(0);
      expect(report.recommendations.length).toBeGreaterThan(0);
    });
    
    test('should provide quality grade and summary', () => {
      const grade = ContentQualityValidator.getQualityGrade(85);
      expect(grade.grade).toBe('B');
      expect(grade.description).toBe('Good quality');
      
      const mockReport = {
        overallScore: 75,
        passesThreshold: true,
        issues: [
          { type: 'warning' as const, message: 'Test warning', severity: 5, category: 'content' as const },
        ],
        recommendations: ['Improve content structure', 'Add more headings'],
        metrics: {
          contentPreservation: 80,
          structureIntegrity: 75,
          markdownQuality: 70,
          readability: 75,
          completeness: 80,
        },
      };
      
      const summary = ContentQualityValidator.generateSummaryReport(mockReport);
      expect(summary).toContain('Quality Report: C');
      expect(summary).toContain('✅ Passes quality threshold');
    });
    
  });
  
  describe('ErrorHandler', () => {
    
    test('should handle retryable errors', async () => {
      const retryableError = new Error('Network timeout occurred');
      
      const result = await ErrorHandler.handleError(retryableError, {
        stage: 'content-extraction',
        operation: 'fetchContent',
        retryCount: 0,
      });
      
      expect(result.fallbackUsed).toBe('retry');
      expect(result.warnings).toContain('Retrying after error: Network timeout occurred');
    });
    
    test('should use fallback strategies for non-retryable errors', async () => {
      const processingError = new Error('Readability parsing failed');

      const result = await ErrorHandler.handleError(processingError, {
        stage: 'readability-processing',
        operation: 'extractContent',
        input: { html: '<p>Test content</p>' },
      });

      expect(['readability-fallback', 'text-only-extraction']).toContain(result.fallbackUsed);
      expect(result.success).toBe(true);
    });
    
    test('should handle memory issues with cleanup strategy', async () => {
      const memoryError = new Error('Out of memory');
      
      const result = await ErrorHandler.handleError(memoryError, {
        stage: 'markdown-conversion',
        operation: 'convertLargeContent',
      });
      
      expect(result.fallbackUsed).toBe('memory-cleanup');
      expect(result.warnings).toContain('Performed memory cleanup, retry recommended');
    });
    
  });
  
  describe('Integration Tests', () => {
    
    test('should process complete workflow without errors', async () => {
      const testHtml = `
        <html>
          <head><title>Test Article</title></head>
          <body>
            <article>
              <h1>Main Title</h1>
              <p>Introduction paragraph with some content.</p>
              
              <h2>Section 1</h2>
              <p>Content for section 1 with <strong>bold text</strong>.</p>
              <ul>
                <li>List item 1</li>
                <li>List item 2</li>
              </ul>
              
              <h2>Section 2</h2>
              <p>More content with <a href="https://example.com">a link</a>.</p>
              <pre><code class="language-javascript">
                console.log('Hello, world!');
              </code></pre>
              
              <h2>Conclusion</h2>
              <p>Final thoughts and summary.</p>
            </article>
          </body>
        </html>
      `;
      
      // Test the complete pipeline
      const result = await OfflineModeManager.processContent(
        testHtml,
        'https://example.com/article',
        'Test Article'
      );
      
      expect(result.success).toBe(true);
      expect(result.markdown).toContain('# Main Title');
      expect(result.markdown).toContain('## Section 1');
      expect(result.markdown).toContain('**bold text**');
      expect(result.markdown).toContain('[a link](https://example.com)');
      expect(result.markdown).toContain('```javascript');
      
      // Validate quality
      const qualityReport = ContentQualityValidator.validate(
        result.markdown,
        testHtml,
        result.processingStats
      );
      
      expect(qualityReport.overallScore).toBeGreaterThan(70);
      expect(qualityReport.passesThreshold).toBe(true);
    });
    
    test('should handle edge cases gracefully', async () => {
      const edgeCases = [
        '', // Empty content
        '<html></html>', // Minimal HTML
        '<script>alert("xss")</script>', // Potentially dangerous content
        'Not HTML at all', // Invalid HTML
      ];
      
      for (const html of edgeCases) {
        const result = await OfflineModeManager.processContent(
          html,
          'https://example.com/edge-case',
          'Edge Case'
        );
        
        // Should not throw errors, even if processing fails
        expect(result).toBeDefined();
        expect(typeof result.success).toBe('boolean');
      }
    });
    
  });
  
});

// Performance benchmarks
describe('Performance Tests', () => {
  
  test('should process medium content within time limits', async () => {
    const mediumContent = '<p>' + 'Content '.repeat(10000) + '</p>';
    const html = `<html><body><article>${mediumContent}</article></body></html>`;
    
    const startTime = performance.now();
    const result = await OfflineModeManager.processContent(
      html,
      'https://example.com/medium',
      'Medium Article'
    );
    const processingTime = performance.now() - startTime;
    
    expect(result.success).toBe(true);
    expect(processingTime).toBeLessThan(5000); // 5 seconds max
  });
  
  test('should handle concurrent processing requests', async () => {
    const html = '<html><body><h1>Test</h1><p>Content</p></body></html>';
    
    const promises = Array.from({ length: 5 }, (_, i) =>
      OfflineModeManager.processContent(
        html,
        `https://example.com/concurrent-${i}`,
        `Concurrent Test ${i}`
      )
    );
    
    const results = await Promise.all(promises);
    
    results.forEach(result => {
      expect(result.success).toBe(true);
      expect(result.markdown).toContain('# Test');
    });
  });
  
});

// Export test utilities for manual testing
export const TestUtils = {
  generateTestHtml: (size: number) => {
    const content = '<p>' + 'Test content '.repeat(size) + '</p>';
    return `<html><body><article>${content}</article></body></html>`;
  },
  
  runManualTest: async (html: string, url: string, title: string) => {
    console.log('Running manual test...');
    const result = await OfflineModeManager.processContent(html, url, title);
    console.log('Result:', result);
    
    const qualityReport = ContentQualityValidator.validate(
      result.markdown,
      html,
      result.processingStats
    );
    console.log('Quality Report:', qualityReport);
    
    return { result, qualityReport };
  },
};
