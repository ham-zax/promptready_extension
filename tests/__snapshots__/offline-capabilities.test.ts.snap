// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Offline Capabilities System > Integration Tests > should process complete workflow without errors 1`] = `
"> Source: https://example.com/article
> Captured: <timestamp>
> Hash: <hash>

# Main Title

## Table of Contents

- [Main Title](#main-title)
- [Section 1](#section-1)
- [Section 2](#section-2)
- [Conclusion](#conclusion)


Introduction paragraph with some content.

## Section 1

Content for section 1 with **bold text**.

- List item 1

- List item 2

## Section 2

More content with [a link](https://example.com/).

\`\`\`javascript

\`\`\`
            console.log('Hello, world!');\`\`\`
\`\`\`

## Conclusion

Final thoughts and summary.
"
`;
